lib/auth/userloginstate/generator.go

```diff
@@ -213,6 +212,16 @@ func (g *Generator) generate(ctx context.Context, user types.User, ulsService se
 
 // addAccessListsToState will add the user's applicable access lists to the user login state after validating them, returning any inherited roles and traits.
 func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) (inheritedRoles []string, inheritedTraits map[string][]string, err error) {
+	locks, err := g.accessLists.GetLocks(ctx, true, types.LockTarget{
+		User: user.GetName(),
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	if len(locks) > 0 {
+		return inheritedRoles, inheritedTraits, nil
+	}
```

Comment:
Is this related to the scalability fix or is it a separate bug fix? If the latter, I'd push this to a separate PR.

Reply:
This change optimizes the scalability fix by reducing redundant lock checks. Previously, GetLocks was invoked multiple times inside `access.IsMembe`r for each edge during traversal, which added unnecessary overhead. I’ve decoupled the lock verification from the hierarchy builder so the check happens once at the appropriate layer.

Where we have regression for this userloginstate.Genrate behavior: 
[name:  "lock prevents adding roles and traits",](https://github.com/gravitational/teleport/blob/master/lib/auth/userloginstate/generator_test.go#L143)